okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt

```diff
@@ -13,430 +13,403 @@
+    call!!.enqueue(object : Callback {
+      override fun onResponse(call: Call, response: Response) {
+        val exchange = response.exchange
+        val streams: Streams
         try {
-          checkUpgradeSuccess(response, exchange);
-          streams = exchange.newWebSocketStreams();
-        } catch (IOException e) {
-          if (exchange != null) exchange.webSocketUpgradeFailed();
-          failWebSocket(e, response);
-          closeQuietly(response);
-          return;
+          checkUpgradeSuccess(response, exchange)
+          streams = exchange!!.newWebSocketStreams()
+        } catch (e: IOException) {
+          exchange?.webSocketUpgradeFailed()
+          failWebSocket(e, response)
+          closeQuietly(response)
+          return
         }
 
         // Process all web socket messages.
         try {
-          String name = "OkHttp WebSocket " + request.url().redact();
-          initReaderAndWriter(name, streams);
-          listener.onOpen(RealWebSocket.this, response);
-          loopReader();
-        } catch (Exception e) {
-          failWebSocket(e, null);
+          val name = "OkHttp WebSocket ${request.url().redact()}"
+          initReaderAndWriter(name, streams)
+          listener.onOpen(this@RealWebSocket, response)
+          loopReader()
+        } catch (e: Exception) {
+          failWebSocket(e, null)
         }
       }
 
-      @Override public void onFailure(Call call, IOException e) {
-        failWebSocket(e, null);
+      override fun onFailure(call: Call, e: IOException) {
+        failWebSocket(e, null)
       }
-    });
+    })
   }
 
-  void checkUpgradeSuccess(Response response, @Nullable Exchange exchange) throws IOException {
+  @Throws(IOException::class)
+  internal fun checkUpgradeSuccess(response: Response, exchange: Exchange?) {
     if (response.code() != 101) {
-      throw new ProtocolException("Expected HTTP 101 response but was '"
-          + response.code() + " " + response.message() + "'");
+      throw ProtocolException(
+          "Expected HTTP 101 response but was '${response.code()} ${response.message()}'")
     }
 
-    String headerConnection = response.header("Connection");
-    if (!"Upgrade".equalsIgnoreCase(headerConnection)) {
-      throw new ProtocolException("Expected 'Connection' header value 'Upgrade' but was '"
-          + headerConnection + "'");
+    val headerConnection = response.header("Connection")
+    if (!"Upgrade".equals(headerConnection, ignoreCase = true)) {
+      throw ProtocolException(
+          "Expected 'Connection' header value 'Upgrade' but was '$headerConnection'")
     }
 
-    String headerUpgrade = response.header("Upgrade");
-    if (!"websocket".equalsIgnoreCase(headerUpgrade)) {
-      throw new ProtocolException(
-          "Expected 'Upgrade' header value 'websocket' but was '" + headerUpgrade + "'");
+    val headerUpgrade = response.header("Upgrade")
+    if (!"websocket".equals(headerUpgrade, ignoreCase = true)) {
```

Comment:
a custom check method seems warranted

`protocolCheck(boolean) { "message" }`

Reply:
Yeah, agreed. I wanna punt to follow-up though.